### PR TITLE
Convert Calico to the ops framework

### PR DESCRIPTION
**Pull request by mateoflorido**
# Summary
This PR introduces the rewrite for Calico, upgrading it from the reactive framework to the operator framework.
## Changes
Here are the key modifications made in this PR:

- **Offloading the Calico node service to Kubernetes**: The Calico node service has been offloaded to Kubernetes, eliminating the need for a separate Linux service to run the node workload.
- **Removal of the node service bits and handling the process during upgrades**: As part of the transition, the PR removes the node service bits from the existing charm. The management of this process during upgrades is now handled within the charm.
- **Addition of ops.lib.manifest to handle the manifests**: Included `ops.lib.manifest` to better handle the manifests in the charm.
- **Switching from etcd envars to Secrets**: The usage of etcd environmental variables has been replaced with Kubernetes Secrets. This change ensures a smooth transition for the `calico/kube-controllers` workload.
- **Creation of an etcd translation laye**r: To ensure compatibility with the operator framework, an etcd reactive translation layer has been created.

### CODE REVIEW

**Comment by addyess** [comment]
**Validation testing**
I noticed somehow the `validate-wheelhouse` tests are still running and somehow passing.  You can take these out now that it's `ops`

**Review by addyess — APPROVED** [review]
Hey Mateo -- hope this helps out with some ideas.

File: lib/charms/kubernetes_libs/v0/etcd.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+"""TODO: Add a proper docstring here.
+
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply lib/charms/kubernetes_libs/v0/etcd.py]: This library was specifically created for this charm. I am currently using it from a local repository, but once the pull request is merged, I will synchronize it with the `kubernetes-libs` repository.

File: lib/charms/operator_libs_linux/v1/systemd.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
+# Copyright 2021 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply lib/charms/operator_libs_linux/v1/systemd.py:1]: Indeed, more [info](https://charmhub.io/operator-libs-linux/libraries/systemd/source-code)

File: lib/charms/kubernetes_libs/v0/etcd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (addyess) [reply lib/charms/kubernetes_libs/v0/etcd.py]: yeah, i can't find that repo [`kubernetes-libs`](https://charmhub.io/kubernetes-libs)

File: upstream/calico/manifests/3.25.1/calico-etcd.yaml [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+      priorityClassName: system-cluster-critical
+      # The controllers must run in the host network namespace so that
+      # it isn't governed by policy that would prevent it from working.
+      hostNetwork: true
+      containers:
+        - name: calico-kube-controllers
+          image: docker.io/calico/kube-controllers:v3.25.1
+          imagePullPolicy: IfNotPresent
+          env:
+            # CK edit: Adding value: null to address the upgrade to ENV from secrets.
```
Comment (Cynerva) [inline upstream/calico/manifests/3.25.1/calico-etcd.yaml:594]: Can you help me understand this? I'm a little confused about what these `value: null` fields do and why they're needed

File: src/calico_manifests.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+        if not json_config:
+            return
+
+        conflist = json.loads(json_config)
+
+        for plugin in conflist.get("plugins"):
+            if plugin.get("type") == "calico":
+                mtu = self.manifests.config.get("mtu")
+                plugin["mtu"] = mtu if mtu else 0
```
Comment (Cynerva) [inline src/calico_manifests.py:37]: I think patching the mtu in the conflist becomes unnecessary if you restore the `__CNI_MTU__` value in the manifest and instead patch the `veth_mtu` value in the ConfigMap. See my other comments about those.

**Review by Cynerva — CHANGES_REQUESTED** [review]
Nice work so far, I like how this is looking overall. I do have some questions and concerns around config options and other things that I think may have been lost in the conversion.

File: metadata.yaml
[the remark this answers is not in the document]
Comment (mateoflorido) [reply metadata.yaml:1]: Yes, you're right. I'm just worried about something: when we remove the resource, will Charmhub simply delete it from the charm page, or will we encounter any issues? If there are no problems, I would like to include the same [approach](https://github.com/charmed-kubernetes/charm-cilium/blob/main/src/charm.py#L193)  we used in Cilium to bundle the `calicoctl` binary in the resources. What do you think?

File: src/charm.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+                log.exception("Failed to get BGPConfiguration")
+                raise e
+
+        ip_mapping = {
+            "bgp-service-cluster-ips": "serviceClusterIPs",
+            "bgp-service-external-ips": "serviceExternalIPs",
+            "bgp-service-loadbalancer-ips": "serviceLoadBalancerIPs",
+        }
+        spec = bgp_config["spec"]
+        spec["asNumber"] = self.config["global-as-number"]
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply src/charm.py:249]: Absolutely, thank you for bringing it to my attention!

File: upstream/calico/manifests/3.25.1/calico-etcd.yaml [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+      "plugins": [
+        {
+          "type": "calico",
+          "log_level": "info",
+          "log_file_path": "/var/log/calico/cni/cni.log",
+          "etcd_endpoints": "__ETCD_ENDPOINTS__",
+          "etcd_key_file": "__ETCD_KEY_FILE__",
+          "etcd_cert_file": "__ETCD_CERT_FILE__",
+          "etcd_ca_cert_file": "__ETCD_CA_CERT_FILE__",
+          "mtu": 10,
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply upstream/calico/manifests/3.25.1/calico-etcd.yaml:91]: This change was necessary to make it a valid JSON and readable in the charm. However, I also believe it's better to let the `install-cni` script do its job and modify it. I've used a workaround to be able to modify the JSON while keeping the `__CNI_MTU__` value intact.

File: upstream/calico/manifests/3.25.1/calico-etcd.yaml (hunk 3 above)
Comment (mateoflorido) [reply upstream/calico/manifests/3.25.1/calico-etcd.yaml:594]: In the reactive version of the charm, we used environment variables with values. However, when we want to apply a new value that now comes from a Secret, the API rejects the manifest, stating that it cannot have both at the same time. This workaround utilizes `value: null` to remove the value and use the one from the Secret instead.

More information: Issue [46861](https://github.com/kubernetes/kubernetes/issues/46861)

File: src/calico_manifests.py (hunk 4 above)
Comment (mateoflorido) [reply src/calico_manifests.py]: Thanks! Removed 👍🏻

File: upstream/calico/manifests/3.25.1/calico-etcd.yaml (hunk 3 above)
Comment (mateoflorido) [reply upstream/calico/manifests/3.25.1/calico-etcd.yaml]: Now the `calico_manifests.py` handle this without modifying the manifest.

File: lib/charms/kubernetes_libs/v0/etcd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (addyess) [reply lib/charms/kubernetes_libs/v0/etcd.py]: is this a library on charmhub you intend to create?

File: build-calico-resource.sh
Comment (Cynerva) [inline build-calico-resource.sh:1]: Now that the calico-node-image resource no longer exists, the `touch calico-node-image.tar.gz` line at the end of this file could be removed too.

File: metadata.yaml
[the remark this answers is not in the document]
Comment (Cynerva) [reply metadata.yaml:1]: When we merge this change, we will have new latest/edge and 1.28/edge builds of the charms that no longer have the resource. But stable channels like 1.27/stable and 1.26/stable should still have them.

Charmhub seems to handle it gracefully as far as I can tell. Here's a test charm with a stable channel that has a resource, and an edge channel in which the resource has been removed from metadata.yaml:

https://charmhub.io/cynerva-test/resources?channel=stable
https://charmhub.io/cynerva-test/resources?channel=edge

[the remark this answers is not in the document]
Comment (Cynerva) [reply metadata.yaml:1]: I think the [resource-spec.yaml](https://github.com/charmed-kubernetes/jenkins/blob/b964af37f8c064606e95057f5df76b575f2af97c/jobs/build-charms/resource-spec.yaml#L10) file in Jenkins will need to be updated though.

**Review by Cynerva — APPROVED** [review]
Looks like all the concerns I raised have been addressed, thanks! I have a couple last nitpicks but no blockers. I would be happy to merge this as-is.

**Review by Cynerva — APPROVED** [review]

**Pull request merged**